### ISSUE TITLE
 Widget got rebuilt due to outdated dependencies

### ISSUE THREAD

**Issue by kezhuw** [flutter/flutter#12992]
Suppose widget `W` depends on `InheritedWidget` `D` based some `condition` be true. After a asynchronous operation, `condition` changed from `true` to `false`. From now on, `W` doesn't depend on `D` actually, but `W` still got rebuilt due to changes in `D`. `condition` can be internal state of `W` or information from other `InheritedWidget`.

I post a test [gist](https://gist.github.com/kezhuw/3ff4cf46d2d60565960caf2257cd1242) to show this.

**Comment by Hixie** [comment]
This was an intentional performance optimization - by not having to clear the dependencies each frame, we can avoid a lot of work.

Is it a problem in practice?

I could imagine a scenario where D keeps changing every frame and W is expensive to rebuild where this could be an issue. To address that case we could provide a flag on InheritedWidget that says "clear the list of dependents each time you notify them", which would make each update more expensive but would reduce the number of "stale" dependents that are tracked.

It would be good to have clear benchmarks based on real code to examine which trade-off is the best overall.

**Comment by kezhuw** [comment]
I catch this by reading the code not in practice coding. Maybe this behavior should be documented, if it is intentional, otherwise clients may be astonished by this behavior and fire similar issues.

Seems that your solution to the imaginary scenario tends to reduce but not eliminate rebuild triggered by stale dependencies, since this "stale" can also be introduced by information changed outside `InheritedWidget`, for example, state change in `StatefulWidget`.

I try to resolve this in #12800 (471efaf7ab946758ba981af7b666c005d2efd69d) by deleting stale dependencies from previous build, but I realized that `BuildContext.inheritFromWidgetOfExactType` is permitted to be called outside current widget's build phase in a later time. Currently, the dependency relationship can be established in three ways, if I am not missing:

* Build phase of this widget, such as `StatelessWidget.build` and `State.build`.
* Build phase of other widget, perhaps ancestor of this widget. Such as `State.didChangeDependencies`.
* Interaction event handlers (e.g. gesture callbacks) or timers. 

In the last two methods, the relationship was established outside build phase of calling `BuildContext`. In that PR, I moved second approach to build phase of the calling `BuildContext` by following opinion I expressed in last paragraph of https://github.com/flutter/flutter/pull/12843#issuecomment-344111398. But I am still in phase diving time to find out why dependency relationship establishment is permitted outside build scope? Without knowledge of this, it should be hard to decide whether to clear dependencies established asynchronously.

In a loose sense, current behavior is correct, but it is definitely not accurate.  With accurate dependency tracking, we can give clients strong guarantee that their widgets never got rebuilt due to stale dependencies. This way widget author has less "Performance considerations" burden in daily coding.

After all, performance is a consideration, client guarantee should also be taken into account.

**Comment by Hixie** [comment]
We can definitely document this better.

The basic rule is that a widget should assume its build method can be called arbitrarily.

I'm reluctant to proactively clear out the dependencies without measuring the costs of both approaches. The current design is very intentional, it's trying to avoid doing much manipulation of the dependency sets, on the assumption that doing so would be more expensive than the rare times you get called unnecessarily.

**Comment by rrousselGit** [comment]
As a comeback to this, what would happen if `InheritedElement.notifyDependent` clear the dependency if it calls `didChangeDependencies` on the dependent?

**Comment by markusaksli-nc** [comment]
Closing as this behavior and the resulting performance considerations are now clearly documented in [StatefulWidget](https://api.flutter.dev/flutter/widgets/StatefulWidget-class.html) and [StatelessWidget](https://api.flutter.dev/flutter/widgets/StatelessWidget-class.html).

**Comment by Hixie** [comment]
@markusaksli-nc Where exactly do you see the discussion of how InheritedWidget can keep rebuilding a Widget that doesn't depend on it anymore in those docs? I don't see it.

**Comment by markusaksli-nc** [comment]
You're right, the **resulting considerations** seemed to be documented clearly, my bad.

> If a widget's parent will regularly change the widget's configuration, or if it depends on inherited widgets that frequently change, then it is important to optimize the performance of the build method to maintain a fluid rendering performance.

> If the widget is likely to get rebuilt frequently due to the use of InheritedWidgets, consider refactoring the stateless widget into multiple widgets, with the parts of the tree that change being pushed to the leaves.

> The second category is widgets that use State.setState or depend on InheritedWidgets. These will typically rebuild many times during the application's lifetime, and it is therefore important to minimize the impact of rebuilding such a widget.

I suppose then that

> The current design is very intentional, it's trying to avoid doing much manipulation of the dependency sets, on the assumption that doing so would be more expensive than the rare times you get called unnecessarily.

this specifically should be clarified somewhere? It's true that "inherited widgets that frequently change" doesn't explicitly address this case.

**Comment by Hixie** [comment]
I think we just need to add a section to InheritedWidget's docs that talks about when dependencies get cleared.

**Comment by rrousselGit** [comment]
Hello @Hixie!

In an effort to unlock features in package:provider & remove the need for subclassing StatelessWidget/ConsumerWidget in package:riverpod, I'd like to solve this issue.

To avoid a breaking change and not negatively impact the performance of widgets that don't care about such thing, what about:

- adding a new `context.dependOn` method which could be used only within `Element.build` 
or:
- adding a flag to `context.dependOnInheritedWidgetOfExactType`, such as `removeDependencyWhenNoLongerUsed: true`


The idea would be to make the behavior of "track dependencies used within `build`" to be opt-in. 

Is that something reasonable?


This change likely would be done in coordination with https://github.com/flutter/flutter/pull/33213 (or alternative implementation).
I'm willing to spend as much time necessary as possible on these issues. They are pretty important for the evolution of Provider/Riverpod

Have a nice day!

**Comment by Hixie** [comment]
@rrousselGit This issue is just about fixing the documentation. I'm not sure I have the context about what you're referencing. Can you file a new bug that describes the problem you'd like to sovle?

**Comment by rrousselGit** [comment]
Oh my bad, I thought this was about changing the described behavior

I'll raise a new issue  then
